packages/unmock-server/src/proxy.ts

```diff
@@ -0,0 +1,95 @@
+import http = require("http");
+import httpProxy = require("http-proxy");
+import net = require("net");
+
+const port = 8008;
+
+const log = (...args: any[]) => console.log(...args); // tslint:disable-line
+
+const TARGET_HOST = "localhost";
+const TARGET_HTTP_PORT = 8000;
+const TARGET_HTTPS_PORT = 8443;
+const TARGET_HTTP_URL = `http://${TARGET_HOST}:${TARGET_HTTP_PORT}`;
+
+const proxy = httpProxy.createProxyServer({});
+
+proxy.on("error", (err, _, res) => {
+  log("proxy error", err);
+  res.end();
+});
+
+proxy.on(
+  "proxyReq",
+  (proxyReq: http.ClientRequest, req: http.IncomingMessage, __, ___) => {
+    proxyReq.setHeader("X-Forwarded-host", req.headers.host || "");
+  },
+);
+
+/**
+ * HTTP proxy. Sets `X-Forwarded-Host` header.
+ */
+const httpServerProxy = (
+  req: http.IncomingMessage,
+  res: http.ServerResponse,
+) => {
+  log("Proxy HTTP request for:", req.url, "sending to", TARGET_HTTP_URL);
+
+  proxy.web(req, res, { target: TARGET_HTTP_URL });
+};
+
+/**
+ * HTTPS proxy
+ */
+const connectListener = (
+  req: http.IncomingMessage,
+  socket: net.Socket,
+  bodyhead: any,
+) => {
+  log(`Request to ${req.headers.host}, ${req.url}, ${req}`);
+
+  const proxySocket = new net.Socket();
+  proxySocket.connect(TARGET_HTTPS_PORT, TARGET_HOST, () => {
+    log(
+      `Connect at ${port} to host ${req.headers.host}, version ${req.httpVersion}, bodyhead: ${bodyhead}`,
+    );
+    proxySocket.write(bodyhead);
+    socket.write(
+      "HTTP/" + req.httpVersion + " 200 Connection established\r\n\r\n",
+    );
+  });
+
+  proxySocket.on("data", chunk => {
+    socket.write(chunk);
+  });
+
+  proxySocket.on("end", () => {
+    socket.end();
+  });
+
+  proxySocket.on("error", () => {
+    socket.write("HTTP/" + req.httpVersion + " 500 Connection error\r\n\r\n");
+    socket.end();
+  });
+
+  socket.on("data", chunk => {
+    proxySocket.write(chunk);
```

Comment:
Do we need any checks on the chunk being written (e.g. that it ends with `\r\n\r\n`)?

Reply:
I'm not sure we need that, it's up to the target server to return proper data. Also not sure how that would work with encryption.